### PR TITLE
Implementation for discrete multisession

### PR DESCRIPTION
**Pull request by introspective-swallow**
I implemented the discrete multisession mode for CEBRA, analogous to the continuous multisession mode. It works for one discrete label and uses the uniform distribution for prior sampling. I also included it in the sklearn integration. I added tests for the sampler, loader, solver and the sklearn integration.

Added objects that differ from the continuous multisession are the loader `DiscreteMultiSessionDataLoader` and the sampler `DiscreteMultisessionSampler`.

As an example, this are session embeddings fitted with discrete multisession for the data I am working with (that will be published soon):

multidisc

I am new to open source contribution, so I'm excited to get your corrections and feedback!

### CODE REVIEW

File: tests/test_sklearn_metrics.py [hunk 1]
```diff
@@ -324,6 +324,67 @@ def test_sklearn_datasets_consistency():
...
+    assert len(datasets) == 4
+ 
+    with pytest.raises(ValueError, match="Labels.*value"):
+        _, _, _ = cebra_sklearn_metrics.consistency_score(
+            [embedding1, embedding2],
+            labels=[np.random.randint(5, size=(10000,)),np.random.randint(10, size=(10000,))],
+            between="datasets")
+    
+
+test_sklearn_datasets_consistency()
```
Comment (stes) [inline tests/test_sklearn_metrics.py:387]:
```suggestion
```

This seems to be left over from testing/prototyping?

File: tests/test_sklearn_metrics.py [hunk 2]
```diff
@@ -324,6 +324,67 @@ def test_sklearn_datasets_consistency():
...
+    )
+    assert scores.shape == (12,)
+    assert pairs.shape == (12, 2)
+    assert len(datasets) == 4
+ 
+    with pytest.raises(ValueError, match="Labels.*value"):
+        _, _, _ = cebra_sklearn_metrics.consistency_score(
+            [embedding1, embedding2],
+            labels=[np.random.randint(5, size=(10000,)),np.random.randint(10, size=(10000,))],
+            between="datasets")
```
Comment (stes) [inline tests/test_sklearn_metrics.py:384]: Because the test is already quite long, it would be better to move this into its own test function.

**Review by stes — COMMENTED** [review]
I added two small remarks, but in general this looks promising. Will start running tests, and then proceed with a more in-depth review.

**Comment by stes** [comment]
Hi @introspective-swallow , thanks a lot for your contribution! I skimmed the code and it looks good overall, made two small comments. I just started to run tests. I'll do another in-depth pass through the code in the next days.

**Comment by stes** [comment]
Also, if you don't mind, do you have an example of a fitted embedding that you could add to the PR description as a reference?

**Comment by stes** [comment]
@introspective-swallow a few tests are failing, could you have a look? I created another PR #135 as some of the failing tests are unrelated to your contribution. I will fix these independently, but others are failing due to the test setup in this PR. Let me know if you have questions about the particular tests.

You can also run them locally using `pytest`!

File: tests/test_sklearn_metrics.py (hunk 1 above)
Comment (introspective-swallow) [reply tests/test_sklearn_metrics.py:387]: Yes, thank you

File: tests/test_sklearn_metrics.py (hunk 2 above)
Comment (introspective-swallow) [reply tests/test_sklearn_metrics.py:384]: This commit was an error of mine, I was implementing the consistency for discrete labels aside and accidentally included it in the PR.

**Comment by introspective-swallow** [comment]
> Also, if you don't mind, do you have an example of a fitted embedding that you could add to the PR description as a reference?

I do have one but the data is private for now though it will be published soon. I could remove the label names and include a fitted embedding, would that work?

**Comment by introspective-swallow** [comment]
> @introspective-swallow a few tests are failing, could you have a look? I created another PR #136 as some of the failing tests are unrelated to your contribution. I will fix these independently, but others are failing due to the test setup in this PR. Let me know if you have questions about the particular tests.
> 
> You can also run them locally using `pytest`!

Yes! I fixed and added some tests and now pytest is giving me no errors.

**Comment by stes** [comment]
Added the changes from #136 to see if tests pass now.

**Comment by stes** [comment]
> Thank you for your contribution. We require contributors to sign our Contributor License Agreement (CLA). We do not have a signed CLA on file for you. In order for us to review and merge your code, please sign our CLA [here](https://forms.gle/o5F7YiLCZ96vfhQ18). After you signed, you can comment on this PR with `@cla-bot check` to trigger another check.

@introspective-swallow could you also fill this form and re-run the CLA check?

**Comment by stes** [comment]
Ok nice, tests look good. Will do another pass through the code in the next days.

**Comment by introspective-swallow** [comment]
> > Thank you for your contribution. We require contributors to sign our Contributor License Agreement (CLA). We do not have a signed CLA on file for you. In order for us to review and merge your code, please sign our CLA [here](https://forms.gle/o5F7YiLCZ96vfhQ18). After you signed, you can comment on this PR with `@cla-bot check` to trigger another check.
> 
> @introspective-swallow could you also fill this form and re-run the CLA check?

@cla-bot check
Yes, here

**Comment by introspective-swallow** [comment]
@cla-bot check

**Comment by introspective-swallow** [comment]
@stes I already filled the CLA and after triggering a check it doesn't seem to work :(

**Comment by stes** [comment]
@introspective-swallow just checked, you filled in the URL to your full github username (https://github.com/introspective-swallow). you need to correct this to only your username, then it will work.

**Comment by stes** [comment]
@cla-bot check

**Comment by stes** [comment]
fixed! cc @introspective-swallow

**Comment by introspective-swallow** [comment]
Thank you! My mistake.

**Comment by stes** [comment]
No worries!

**Review by MMathisLab — APPROVED** [review]
lgtm; I did not run it though; just overall code review. @stes can final merge confirm here

**Comment by MMathisLab** [comment]
only last issue is to edit the docs to say how to use it. thanks again @introspective-swallow -- sorry for the delay in reviewing this!

File: cebra/datasets/demo.py [hunk 3]
```diff
@@ -113,13 +113,17 @@ def discrete_index(self):
 
 # TODO(stes) remove this from the demo datasets until multi-session training
 # with discrete indices is implemented in the sklearn API.
```
Comment (stes) [inline cebra/datasets/demo.py:115]:
```suggestion
```

no longer needed

**Review by MMathisLab — APPROVED** [review]
lgtm! thanks again for this contribution 💪

**Comment by MMathisLab** [comment]
one minor doc issue to fix (I can do so tomorrow!):

```bash

_____________________________ [doctest] usage.rst ______________________________
1039                                                       num_batches=5)
1040 
1041     # multi-session
1042     multi_score = cebra.sklearn.metrics.infonce_loss(multi_cebra_model,
1043                                                      neural_session1,
1044                                                      continuous_label1,
1045                                                      session_id=0,
1046                                                      num_batches=5)
1047 
1048 
UNEXPECTED EXCEPTION: ValueError('Labels invalid: must have the same type of features as the ones used for fitting,expects int64, got float64.')
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/pytest_sphinx.py", line 366, in _DocTestRunner__run
    exec(
  File "<doctest usage.rst[33]>", line 9, in <module>
  File "/home/runner/work/CEBRA/CEBRA/cebra/integrations/sklearn/metrics.py", line 87, in infonce_loss
    cebra_model._check_labels_types(y, session_id=session_id)
  File "/home/runner/work/CEBRA/CEBRA/cebra/integrations/sklearn/cebra.py", line [86](https://github.com/AdaptiveMotorControlLab/CEBRA/actions/runs/9258095298/job/25467946597?pr=135#step:10:87)3, in _check_labels_types
    raise ValueError(
ValueError: Labels invalid: must have the same type of features as the ones used for fitting,expects int64, got float64.
usage.rst:1048: UnexpectedException
```

**Review by MMathisLab — APPROVED** [review]

**Comment by MMathisLab** [comment]
will will fix the rest of the issues in this issue, as it's related to matplotlib and not this PR: https://github.com/AdaptiveMotorControlLab/CEBRA/pull/150

**Pull request merged**